Firebase/Messaging/FIRMessagingAnalytics.m

```diff
@@ -0,0 +1,230 @@
+
+// Data Key
+static NSString *const kDataKey = @"data";
+
+// Messaging From Key
+static NSString *const kFIRMessagingFromKey = @"from";
+
+static NSString *const kFIRParameterLabel = @"label";
+
+static NSString *const kReengagementSource = @"Firebase";
+static NSString *const kReengagementMedium = @"notification";
+
+// Analytics
+static NSString *const kAnalyticsEnabled =              @"google.c.a." @"e";
+static NSString *const kAnalyticsComposerIdentifier =   @"google.c.a." @"c_id";
+static NSString *const kAnalyticsComposerLabel =        @"google.c.a." @"c_l";
+static NSString *const kAnalyticsMessageLabel =         @"google.c.a." @"m_l";
+static NSString *const kAnalyticsMessageTimestamp =     @"google.c.a." @"ts";
+static NSString *const kAnalyticsMessageUseDeviceTime = @"google.c.a." @"udt";
+static NSString *const kAnalyticsTrackConversions =     @"google.c.a." @"tc";
+
+@implementation FIRMessagingAnalytics
+
++ (BOOL)canLogNotification:(NSDictionary *)notification {
+  if (!notification.count) {
+    // Payload is empty
+    return NO;
+  }
+  NSString *isAnalyticsLoggingEnabled = notification[kAnalyticsEnabled];
+  if (![isAnalyticsLoggingEnabled isKindOfClass:[NSString class]] ||
+      ![isAnalyticsLoggingEnabled isEqualToString:@"1"]) {
+    // Analytics logging is not enabled
+    FIRMessagingLoggerDebug(kFIRMessagingMessageCodeAnalytics001,
+                            @"Analytics logging is disabled. Do not log event.");
+    return NO;
+  }
+  return YES;
+}
+
++ (void)logOpenNotification:(NSDictionary *)notification
+                toAnalytics:(id<FIRAnalyticsInterop> _Nullable)analytics {
+  [self logUserPropertyForConversionTracking:notification toAnalytics:analytics];
+  [self logEvent:kFIRIEventNotificationOpen
+withNotification:notification
+     toAnalytics:analytics];
+}
+
++ (void)logForegroundNotification:(NSDictionary *)notification
+                      toAnalytics:(id<FIRAnalyticsInterop> _Nullable)analytics {
+  [self logEvent:kFIRIEventNotificationForeground
+withNotification:notification
+     toAnalytics:analytics];
+}
+
++ (void)logEvent:(NSString *)event
+withNotification:(NSDictionary *)notification
+     toAnalytics:(id<FIRAnalyticsInterop> _Nullable)analytics {
+  if (!event.length) {
+    FIRMessagingLoggerDebug(kFIRMessagingMessageCodeAnalyticsInvalidEvent,
+                             @"Can't log analytics with empty event.");
+    return;
+  }
+  NSMutableDictionary *params = [self paramsForEvent:event withNotification:notification];
+
+  [analytics logEventWithOrigin:@"fcm" name:event parameters:params];
+  FIRMessagingLoggerDebug(kFIRMessagingMessageCodeAnalytics005,
+                           @"%@: Sending event: %@ params: %@", kLogTag, event, params);
+}
+
++ (NSMutableDictionary *)paramsForEvent:(NSString *)event
+                       withNotification:(NSDictionary *)notification {
+  NSDictionary *analyticsDataMap = notification;
```

Comment:
Nit: Any particular reason why this is renamed here and not other places? Not an issue really, just checking.

Reply:
Only historical.  This code is mostly unchanged from being moved to InstanceID and back here.